File: src/test/scala/workshop/part3/SuperComputeActorTest.scala
package workshop.part3

import scala.language.postfixOps
import akka.actor._
import scala.concurrent.duration._
import workshop.AkkaSpec
import workshop.work.{RiskyWorkException, RiskyWork, RiskyAdditionResult, RiskyAddition}


class SuperComputeActorTest extends AkkaSpec {
  val timeout: FiniteDuration = 300 millis

  it should "compute risky work when work has no failures" in  {
    suppressStackTraceNoise {
      val superComputeActor = system.actorOf(Props(classOf[SuperComputeActor]))

      superComputeActor ! RiskyAddition(1,3,0)
      expectMsg(RiskyAdditionResult(4))
    }
  }

  it should "compute risky work in parallel when work has no failures" in  {
    suppressStackTraceNoise {
      val superComputeActor = system.actorOf(Props(classOf[SuperComputeActor]))

      val workList = List(RiskyAddition(1,3,150), RiskyAddition(2,3,150),  RiskyAddition(4,2,150))
      val workListResults = List(RiskyAdditionResult(4),RiskyAdditionResult(5),RiskyAdditionResult(6))

      workList.foreach(
        w => superComputeActor ! w
      )

      expectParallel(400) {
        val result1 = expectMsgClass(timeout, classOf[RiskyAdditionResult])
        val result2 = expectMsgClass(timeout, classOf[RiskyAdditionResult])
        val result3 = expectMsgClass(timeout, classOf[RiskyAdditionResult])
        List(result1, result2, result3).foreach(workListResults should contain (_))
      }
    }
  }

  it should "compute risky work in parallel when work has failures" in  {
    suppressStackTraceNoise {
      class WorkWithFailure extends RiskyWork {
        override def perform() = throw new RiskyWorkException("test exception")
      }
      val superComputeActor = system.actorOf(Props(classOf[SuperComputeActor]))

      val workList = List(RiskyAddition(1, 3, 150), new WorkWithFailure(),  RiskyAddition(4,2,150))
      val workListResults = List(RiskyAdditionResult(4),RiskyAdditionResult(6))

      workList.foreach(
        w => superComputeActor ! w
      )

      expectParallel(290) {
        val result1 = expectMsgClass(timeout, classOf[RiskyAdditionResult])
        val result2 = expectMsgClass(timeout, classOf[RiskyAdditionResult])
        List(result1, result2).foreach(workListResults should contain (_))
      }
    }
  }
}

File: src/test/java/workshop/part3/SuperComputeActorTest.java
package workshop.part3;

import akka.actor.ActorRef;
import akka.actor.Props;
import org.junit.Test;
import workshop.AkkaTest;
import workshop.work.RiskyWorkException;
import workshop.work.Work;

import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static workshop.helpers.AkkaTestHelper.getGivenNumberOfResultsWithin;
import static workshop.helpers.AkkaTestHelper.getResult;
import static workshop.work.Work.*;

public class SuperComputeActorTest extends AkkaTest {

    @Test
    public void shouldComputeRiskyWorkWhenWorksHasNoFailures() {
        ActorRef superComputeActor = createSuperComputeActor();

        superComputeActor.tell(new RiskyAddition(1, 3), probe.ref());
        RiskyAdditionResult result = getResult(probe, RiskyAdditionResult.class);
        assertThat(result, is(new RiskyAdditionResult(4)));
    }

    @Test
    public void shouldComputeRiskyWorkInParallelWhenWorkHasNoFailures() {
		ActorRef superComputeActor = createSuperComputeActor();

        List<RiskyWork> workList = Arrays.asList(new RiskyAddition(1, 3, 150), new RiskyAddition(2, 3, 150), new RiskyAddition(4, 2, 150));
        List<RiskyAdditionResult> workListResults = new LinkedList<>(Arrays.asList(new RiskyAdditionResult(4), new RiskyAdditionResult(5), new RiskyAdditionResult(6)));

        workList.forEach(work -> superComputeActor.tell(work, probe.ref()));

        List<RiskyAdditionResult> results = getGivenNumberOfResultsWithin(probe, workListResults.size(), 200, RiskyAdditionResult.class);
        results.forEach(result -> assertResult(result, workListResults));

		assertTrue(workListResults.isEmpty());
    }

    @Test
    public void shouldComputeRiskyWorkInParallelWhenWorkHasFailures() {
        class WorkWithFailure extends RiskyWork {
            @Override
            public Work.RiskyWorkResult perform() throws RiskyWorkException {
                RiskyWorkException riskyWorkException = new RiskyWorkException("Risky work failed");
                riskyWorkException.setStackTrace(new StackTraceElement[]{});
                throw riskyWorkException;
            }
        }

        ActorRef superComputeActor = createSuperComputeActor();

        List<RiskyWork> workList = Arrays.asList(new RiskyAddition(1, 3, 150), new WorkWithFailure(), new RiskyAddition(4, 2, 150));
        List<RiskyAdditionResult> workListResults = new LinkedList<>(Arrays.asList(new RiskyAdditionResult(4), new RiskyAdditionResult(6)));

        workList.forEach(work -> superComputeActor.tell(work, probe.ref()));

        List<RiskyAdditionResult> results = getGivenNumberOfResultsWithin(probe, workListResults.size(), 200, RiskyAdditionResult.class);
		results.forEach(result -> assertResult(result, workListResults));
    }

    private ActorRef createSuperComputeActor() {
        return system.actorOf(Props.create(SuperComputeActor.class));
    }

	private void assertResult(final RiskyAdditionResult result, List<RiskyAdditionResult> expectedResults) {
		assertTrue("Unexpected result [" + result + "] expected one of " + expectedResults, expectedResults.remove(result));
	}
}
